### DISCUSSION TITLE
Funding early stage of code review guild

### DISCUSSION THREAD

**Discussion by yangwao** [kodadot/nft-gallery#2999] [General]
Hey, I guess @kodadot/code-review-guild doing a vital part of our way to push stuff to production on the scale to distribute load off one's head, back then me & @vikiival and adding quality for code reviews and limit the damage.

As we value your time and attention, we are looking for ways to put the right rewards and incentives to sustainably help us grow the codebase. 

Yet briefly chatted w @petersopko we will try to drive some numbers from PRs actions by code reviewers as my earliest promise to figure out these things what's would be the best fit-out.

Current numbers, we had running Code Review Guild since approx 2022/04/04, which resulted in 102 merged PRs. 
The early set of folks was 4, so that's approx 25.5 as we've been using round robin and not load balance, it's just average for the easiest approximation.

![image](https://user-images.githubusercontent.com/5887929/167823229-c3a6e5ee-e853-447c-ab08-7e6b53eaa068.png)

Right now are thinking of rewarding $500-$1k, which could be in the range of $20-$40 per code review.
Yet this has some bonkers. If the pull request is not complex it doesn't make much sense upper bound, hence when the issue is complex and even then code review is made and some bug appeared afterwards, which doesn't fix all things. 

Namely participants
- @roiLeo 
- @prachi00 
- @kkukelka 
- @vikiival 
- @petersopko (added recently)

### Ref 
- #2730 
- https://github.com/kodadot/nft-gallery/pulls?q=is%3Apr+is%3Amerged+merged%3A%3E%3D2022-04-04+

**Comment by roiLeo** [comment]
> The round robin algorithm chooses reviewers based on who's received the least recent review request, focusing on alternating between all members of the team regardless of the number of outstanding reviews they currently have.

> The load balance algorithm chooses reviewers based on each member's total number of recent review requests and considers the number of outstanding reviews for each member. The load balance algorithm tries to ensure that each team member reviews an equal number of pull requests in any 30 day period.

What are we on right now? 
Why not just asign @kodadot/code-review-guild? and members will take or suggest someone to review.

> Yet this has some bonkers. If the pull request is not complex it doesn't make much sense upper bound, hence when the issue is complex and even then code review is made and some bug appeared afterwards, which doesn't fix all things.

It may depend on the reward of the issue (small bounty, small reward)

**Reply by kkukelka** [reply]
> > Yet this has some bonkers. If the pull request is not complex it doesn't make much sense upper bound, hence when the issue is complex and even then code review is made and some bug appeared afterwards, which doesn't fix all things.
> 
> It may depend on the reward of the issue (small bounty, small reward)

this ☝️

**Reply by yangwao** [reply]
> What are we on right now?

> The early set of folks was 4, so that's approx 25.5 as we've been using round robin and not load balance,

Moved from round-robin to load balance.

> Why not just asign https://github.com/orgs/kodadot/teams/code-review-guild? and members will take or suggest someone to review.

Because you want to rotate between sub-set of folks in a group to everyone grant equal choice have exposure to particular code and save willpower on other things, where algo can take care of this.

I mean, if the wrong member is assigned, I've noticed in some cases (<20%) adding someone extra helped or the original reviewer calls someone else.

**Comment by yangwao** [comment]
> It may depend on the reward of the issue (small bounty, small reward)

Yet it would require writing some system for it to establish. 
Which could be part of Assing bot v3 I guess or we will see what would work best. 

@petersopko been sharing numbers from PRs review activty tho, so it's time to discuss it :)

**Comment by petersopko** [comment]
these are the current numbers when it comes to reviews, `prReviewCount` is how many PRs did somebody review, `prReviewInteractionCount` is how many interactions happened in total from said dev (interactions such as `APPROVED`, `CHANGES_REQUESTED`, `COMMENTED`)
edit: updated 15.06.2022

```json
{
  "roiLeo": {
    "prReviewCount": 217,
    "prReviewInteractionCount": 390
  },
  "vikiival": {
    "prReviewCount": 194,
    "prReviewInteractionCount": 336
  },
  "kkukelka": {
    "prReviewCount": 47,
    "prReviewInteractionCount": 58
  },
  "yangwao": {
    "prReviewCount": 33,
    "prReviewInteractionCount": 40
  },
  "prachi00": {
    "prReviewCount": 21,
    "prReviewInteractionCount": 25
  },
  "petersopko": {
    "prReviewCount": 19,
    "prReviewInteractionCount": 22
  },
  "Jarsen136": {
    "prReviewCount": 1,
    "prReviewInteractionCount": 1
  },
  "KngZhi": {
    "prReviewCount": 1,
    "prReviewInteractionCount": 1
  },
  "marsic3": {
    "prReviewCount": 1,
    "prReviewInteractionCount": 2
  },
  "dezine2dev": {
    "prReviewCount": 1,
    "prReviewInteractionCount": 1
  }
}


```

**Reply by kkukelka** [reply]
damn, @roiLeo is a beast 💪

**Comment by kkukelka** [comment]
Any updates in this regard? Would be a helpful incentive for some of us, as I currently only have tome to review some PRs throughout the day.

**Reply by petersopko** [reply]
updated table

**Comment by yangwao** [comment]
> we had been running Code Review Guild since approx 2022/04/04

can we have numbers since this date @petersopko ?

**Reply by petersopko** [reply]
here are the updated numbers, closed/merged PRs reviewed >= 2022-04-04 (where author of PR isn't reviewer)
```json
{
  "roiLeo": {
    "prReviewCount": 53,
    "prReviewInteractionCount": 113
  },
  "vikiival": {
    "prReviewCount": 46,
    "prReviewInteractionCount": 97
  },
  "kkukelka": {
    "prReviewCount": 37,
    "prReviewInteractionCount": 45
  },
  "prachi00": {
    "prReviewCount": 22,
    "prReviewInteractionCount": 26
  },
  "petersopko": {
    "prReviewCount": 19,
    "prReviewInteractionCount": 22
  },
  "KngZhi": {
    "prReviewCount": 1,
    "prReviewInteractionCount": 1
  },
  "Jarsen136": {
    "prReviewCount": 1,
    "prReviewInteractionCount": 1
  },
  "yangwao": {
    "prReviewCount": 1,
    "prReviewInteractionCount": 1
  }
}
```

**Comment by yangwao** [comment]
I'm thinking aloud about drafting the first proposal till we "deploy" the right system to keep things going as we have a quite loaded roadmap. 

As usual, it's the first shot, so feel free to have objections, I also don't want to introduce wrong metrics, it's all about experimenting here and rewarding for your long-term participation. 

I guess let's take roiLeo as an example. Assigned 53 PRs and 113 interactions (comments, code reviews (?), &c). If we go averaging as some PRs might be super light, where you spent less than 20 minutes, at some might hit up to an hour, going with $30 accounting for context switching should be a sweet deal. 

The formula could be `(number of PRs assigned + interactions * appraisal constant ) * average value of code review`
 Appraisal constant is something we might play around with. For example, if one accrues more appraisal from the audience for their interactions, would be higher and is in the charge of a prior audience of a particular PR. `0.2` is just a head start to not game it just for sake of comment/bugging around and also encourage to interact in fair balance, as some interactions might be not worth time and might slow down things for merging PR

Till we figure out a better system, we can think of making it on some n-weeks basis, like 4~8-weeks 

The brief calculation in particular order would be
`(53+113*0.2)*30 = 2268`
`(46+97*0.2)*30 = 1962` 
`(37+45*0.2)*30 = 1380`
`(22+26*0.2)*30 = 816` 
`(19+22*0.2)*30 = 702`

Noting down addresses in a particular order, took them from your last payout PRs, I'll proceed with payouts soon

- https://beta.kodadot.xyz/transfer/?target=HguGafrfk8nRP9a4tgJf7hLvSSFie8GtkDjyWFHQ2kDUAMc
  - https://kusama.subscan.io/extrinsic/0x059ef6ff269a3df57153b6d26baddf102f64229e245e388dbc361c608ca4750d

- https://beta.kodadot.xyz/transfer/?target=Fksmad33PFxhrQXNYPPJozgWrv82zuFLvXK7Rh8m1xQhe98
  - https://kusama.subscan.io/extrinsic/0x45f9814f8f6f70b78c80c49a4dd01614109589903210d21072b6d90dd6d7c8d0
- https://beta.kodadot.xyz/transfer/?target=EqdyzrzVmeHwMdMwvPeCMnNdbuQDbD3YrjY93xq9Ln3jUGW
  - https://kusama.subscan.io/extrinsic/0x3cb80a4839856ffcebbe0aaf24e592709539f5e3b90e61844c6c41d7208d0f3c
- https://beta.kodadot.xyz/transfer/?target=EzGc4s9PgCPx1YnF3fqzhLzVHpHMTL4LWPScwpDrR8JKgSU
  - https://kusama.subscan.io/extrinsic/0xd0c89244ad6230ee759b02376c1d41bf993c6c471435aa6f5718b67279ebf66f
- https://beta.kodadot.xyz/transfer/?target=FNyt7T1xdbhN7dagf7yGYCRJuE3R45VmTCE3tjzy1rKxa7y
  - https://kusama.subscan.io/extrinsic/0xdeccf7e46c8520fdf30273437b94778db9aec9eb4e947bc2fe926f98b53bc901

For the future, seems @Jarsen136 would be eligible to help us in the code review guild if you would want to! :)

**Reply by kkukelka** [reply]
Yes @Jarsen136 really provided so much value to the project in the past months. As he has acquired a pretty good understanding of the repository it would be great to have him onboard in the guild!

**Reply by Jarsen136** [reply]
Thank you, guys! I am willing to join in code review guild. 😄

**Reply by yangwao** [reply]
Let's accept invite and welcome 🙌 😄
<img width="246" alt="image" src="https://user-images.githubusercontent.com/5887929/174312426-eec5f5d9-020c-45b2-81f6-229d07ad7dc9.png">

**Comment by yangwao** [comment]
Sent $7128 in $KSM, enjoy! :)
Let's see if it's a working model from here.
Hope it helps adds code quality, avoid adding too much technical debt and of course working stuff :)
